merkledag/merkledag.go

```diff
@@ -146,21 +145,58 @@ func FindLinks(links []key.Key, k key.Key, start int) []int {
 	return out
 }
 
+func (ds *dagService) GetMany(ctx context.Context, keys []key.Key) (<-chan *Node, <-chan error) {
+	out := make(chan *Node)
+	errs := make(chan error, 1)
+	blocks := ds.Blocks.GetBlocks(ctx, keys)
+	var count int
+
+	go func() {
+		defer close(out)
+		defer close(errs)
+		for {
+			select {
+			case b, ok := <-blocks:
+				if !ok {
+					if count != len(keys) {
+						errs <- fmt.Errorf("failed to fetch all nodes")
+					}
+					return
+				}
+				nd, err := Decoded(b.Data)
+				if err != nil {
+					errs <- err
+					return
+				}
+				select {
+				case out <- nd:
+					count++
+				case <-ctx.Done():
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return out, errs
+}
+
 // GetDAG will fill out all of the links of the given Node.
 // It returns a channel of nodes, which the caller can receive
 // all the child nodes of 'root' on, in proper order.
-func (ds *dagService) GetDAG(ctx context.Context, root *Node) []NodeGetter {
+func GetDAG(ctx context.Context, ds DAGService, root *Node) []NodeGetter {
```

Comment:
Ah ha, I grok this now: minimize the interface surface area where ever possible. That makes excellent sense -- I didn't realize how I was coming at this from a much more Java classes perspective. +:100: